subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultImmutableAttributes.java

```diff
@@ -206,6 +206,9 @@ public Object get() {
         if (attributeType.isAssignableFrom(attribute.getType())) {
             return (S) get();
         }
+        if (String.class == attributeType) {
+            return Cast.uncheckedCast(value.isolate().toString());
+        }
```

Comment:
This is a coercion to `String`, as to coerce `Usage` to a string. Unfortunately, I have nothing better than this, but I'm happy to consider alternatives.

Reply:
When do you need to do this coercion? Just to 'desugar' attributes for reporting? If so, I'd expect this to be explicit, not implicit. In fact, maybe a 'desugared' attribute container is no longer an attribute container, but just a `Map<String, String>`.

It would be great if you could talk to @adammurdoch about how attributes work, so that we have a clear plan for evolving them.

Reply:
No, the coercion is because of merging the component attributes and variant attributes that this PR does. Without that, there's a flaky failure, depending on the order of which the attributes show up.

Reply:
(note that desugaring is for the reporting PR, which has nothing to do with this one)

Reply:
OK I'm going to give up on trying to understand why our attributes work how they do. I'm going to leave it to you and Adam, and assume that you have a common understanding of how we want to evolve them.
I'll leave it up to you to make sure that you _do_ have that common understanding.

Reply:
To be clear: this is the only solution I have found to make this work _consistently_. The more I look at this, the more I think we should get rid of the typed attributes, as it adds a lot of complexity, and I just _don't_ understand why typically here I can reach such a situation:

![selection_118](https://user-images.githubusercontent.com/316357/34362176-af698dbe-ea71-11e7-8261-bdae927a6f87.png)